mlflow/store/ftp_artifact_repo.py

```diff
@@ -0,0 +1,124 @@
+
+        super(FTPArtifactRepository, self).__init__(artifact_uri)
+
+    def _is_dir(self, full_file_path):
+        try:
+            self.ftp.cwd(full_file_path)
+            return True
+        except ftplib.error_perm:
+            return False
+
+    def _mkdir(self, artifact_dir):
+        try:
+            self.ftp.mkd(artifact_dir)
+        except ftplib.error_perm:
+            head, _ = os.path.split(artifact_dir)
+            self._mkdir(head)
+            self._mkdir(artifact_dir)
+
+    def download_files(self, path, destination):
+        self.ftp.cwd(path)
+        if not os.path.isdir(destination):
+            os.makedirs(destination)
+
+        filelist = self.ftp.nlst()
+
+        for ftp_file in filelist:
+            if self._is_dir(build_path(path, ftp_file)):
+                self.download_files(build_path(path, ftp_file), build_path(destination, ftp_file))
+            else:
+                with open(os.path.join(destination, ftp_file), "wb") as f:
+                    self.ftp.retrbinary("RETR "+ftp_file, f)
+        return
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_dir = os.path.join(self.path, artifact_path) \
+            if artifact_path else self.path
+        self._mkdir(artifact_dir)
+        with open(local_file, 'rb') as f:
+            self.ftp.cwd(artifact_dir)
+            self.ftp.storbinary('STOR ' + os.path.split(local_file)[1], f)
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        dest_path = os.path.join(self.path, artifact_path) \
+            if artifact_path else self.path
+
+        dest_path = build_path(dest_path, os.path.split(local_dir)[1])
+        dest_path_re = os.path.split(local_dir)[1]
+        if artifact_path:
+            dest_path_re = build_path(artifact_path, os.path.split(local_dir)[1])
+
+        local_dir = os.path.abspath(local_dir)
+        for (root, _, filenames) in os.walk(local_dir):
+            upload_path = dest_path
+            if root != local_dir:
+                rel_path = get_relative_path(local_dir, root)
+                upload_path = build_path(dest_path_re, rel_path)
+            if not filenames:
+                self._mkdir(build_path(self.path, upload_path))
+            for f in filenames:
+                if os.path.isfile(build_path(root, f)):
+                    self.log_artifact(build_path(root, f), upload_path)
+
+    def list_artifacts(self, path=None):
+        artifact_dir = self.path
+        list_dir = os.path.join(artifact_dir, path) if path else artifact_dir
+        artifact_files = self.ftp.nlst(list_dir)
+        infos = []
+        for file_name in artifact_files:
+            file_path = file_name if path is None else os.path.join(path, file_name)
+            full_file_path = os.path.join(list_dir, file_name)
+            if self._is_dir(full_file_path):
+                infos.append(FileInfo(file_path, True, None))
+            else:
+                self.ftp.voidcmd('TYPE I')
```

Comment:
Q: why is this (setting the transfer mode to binary) necessary / why do we perform it here? It seems like a side-effecting command, so would be good to avoid if possible.

Reply:
without transfer mode to binary will cause `ftp.size` failure. I have no idea ti avoid it.